### PR TITLE
Android plugin code refactoring

### PR DESCRIPTION
**Pull request by vincent-unity**
This pr mainly did the Android plugin code refactoring, including:
1. Reorder the functions in `UnityNotificationManager` by grouping them with functionalities.
2. Only make APIs public when necessary, change others to protected if it's only used internally.
3. Removed `UnityNotificationManager.getScheduledNotifications()` and its related functions as they're not used for now, we can add them back when we really want to use.
4. Move some APIs to `UnityNotificationManagerUtilities`, like `serializeNotificationIntent/deserializeNotificationIntent` etc.
5. Rename some functions to java style, also format the files to follow java convention.
6. Add simple comment to the funtions, also some code snippets.
7. Use `name` rather than `title` for NotificationChannel, as `name` is what Android is using. 
8. Other minor changes

To QA: Please help to verify if it on 2018.4/2019.3 as well, also please verify on device lower than Android N, as I don't have any Android 6 or below devices.

### CODE REVIEW

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java [hunk 1]
```diff
@@ -2,64 +2,50 @@
...
-        Notification.Builder notificationBuilder = UnityNotificationManager.buildNotification(intent, context);
-        int id = intent.getIntExtra("id", -1);
+    // Send notification for Android N and above.
+    protected static void sendNotificationNougat(Context context, Intent intent) {
+        Notification.Builder notificationBuilder = UnityNotificationManager.buildNotification(context, intent);
 
+        // TODO: setGroup/setGroupSummary/setSortKey are introduced in API Level 20, why we check N here?
+        // https://developer.android.com/reference/android/app/Notification.Builder
+        // If this is something we should move to UnityNotificationManager.buildNotification(), we probably can remove
+        // UnityNotificationManagerNougat as we don't need to do something special for Android N.
```
Comment (vincent-unity) [inline com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java]: @todi1856 @aurimasc This is something I want to discuss with you guys. 
`setGroup/setGroupSummary/setSortKey` are introduced in API Level 20, **_why we check N here?_**  Do you guys have any idea? Here is the document from google: https://developer.android.com/reference/android/app/Notification.Builder
If this is something we should move to UnityNotificationManager.buildNotification(), we probably can remove
UnityNotificationManagerNougat as we don't need to do something special for Android N. 
How do you think?

**Comment by vincent-unity** [comment]
@lucianUnity3D FYI.

**Comment by todi1856** [comment]
I just noticed that AndroidNotificationResourcesPostProcessor.cs will fail to find build.gradle if building with Visual Studio export, the path is done 

var gradleFilePath = Path.Combine(projectPath, "build.gradle");

but for VS, it should be var gradleFilePath = Path.Combine(projectPath, "build.gradle");

could you change the code to 

            var gradleFilePath = Path.Combine(projectPath, PlayerSettings.productName + "/build.gradle");
            if (!File.Exists(gradleFilePath))
                gradleFilePath = Path.Combine(projectPath, "build.gradle");
            if (!File.Exists(gradleFilePath))
                throw new FileNotFoundException(string.Format("'{0}' doesn't exist.", gradleFilePath));

so it would work both Gradle & VS

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java [hunk 2]
```diff
@@ -2,79 +2,33 @@
...
             boolean canBypassDnd,
             boolean canShowBadge,
             long[] vibrationPattern,
-            int lockscreenVisibility)
-    {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
-        {
-            NotificationChannel channel = new NotificationChannel(id, title, importance);
+            int lockscreenVisibility) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
Comment (todi1856) [inline com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java]: unneeded condition ?

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java (hunk 1 above)
Comment (todi1856) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java]: Just finished at the full code, and I think you're right should probably be removed , especially due this code

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                UnityNotificationManagerNougat.sendNotificationNougat(context, intent);
            } else {
                UnityNotificationManager.sendNotification(context, intent);
            }

Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java]: yeah, we're checking `setGroup/setGroupSummary/setSortKey` with the wrong version code, so nothing special for Android N.
I'll remove `UnityNotificationManagerNougat` and move some check to `UnityNotificationManagerOreo`.

**Comment by vincent-unity** [comment]
Talked with Tomas offline, we agreed we should remove UnityNotificationManagerNougat, and move some code into UnityNotificationManager & UnityNotificationManagerOreo. 
In order to make the changes more trackable, I planned to do it in a separate pr.

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java (hunk 1 above)
Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerNougat.java]: Talked with Tomas offline, we agreed we should remove UnityNotificationManagerNougat, and move some code into UnityNotificationManager & UnityNotificationManagerOreo.
In order to make the changes more trackable, I planned to do it in a separate pr.

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java [hunk 3]
```diff
@@ -87,30 +41,55 @@ public void registerNotificationChannel(
...
+                return notificationChannelToWrapper(ch);
         }
+        return null;
+    }
 
-        return channelList.toArray(new NotificationChannelWrapper[channelList.size()]);
+    protected static NotificationChannelWrapper notificationChannelToWrapper(NotificationChannel channel) {
+        NotificationChannelWrapper wrapper = new NotificationChannelWrapper();
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
[the remark this answers is not in the document]
Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java]: Discussed with Tomas offline, will clean this up in a separate pr to make changes more trackable.

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java (hunk 2 above)
Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManagerOreo.java]: Discussed with Tomas offline, will clean this up in a separate pr to make changes more trackable.

File: com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManager.java [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+        }
+    }
+
+    // Get all notification ids from SharedPreferences.
+    protected int[] getScheduledNotificationIDs() {
+        SharedPreferences prefs = mContext.getSharedPreferences(NOTIFICATION_IDS_SHARED_PREFS, Context.MODE_PRIVATE);
+        Set<String> ids = prefs.getStringSet(NOTIFICATION_IDS_SHARED_PREFS_KEY, new HashSet<String>());
+
+        // Convert the string array ids to int array ids.
+        String[] idsArrStr = ids.toArray(new String[ids.size()]);
```
[the remark this answers is not in the document]
Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManager.java]: Not sure...
Nice catch anyway, I'll do that in a separator pr along with the UnityNotificationManagerNougat refactor to make the changes more trackable.

[the remark this answers is not in the document]
Comment (vincent-unity) [reply com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationManager.java]: Fixed in https://github.com/Unity-Technologies/com.unity.mobile.notifications/pull/13.

**Pull request merged**